File: .submodules/nostos/internal/cluster/bootstrap.go
package cluster

import (
	"context"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/yurifrl/nostos/internal/config"
	"github.com/yurifrl/nostos/internal/paths"
	"github.com/yurifrl/nostos/internal/secrets"
	"gopkg.in/yaml.v3"
)

// Bootstrap runs `talosctl bootstrap` against node. Idempotent on already-bootstrapped.
func Bootstrap(ctx context.Context, cfg *config.Config, p paths.Paths, node config.Node, timeout time.Duration) error {
	if err := requireTalosctl(); err != nil {
		return err
	}
	if node.Role != "controlplane" {
		return fmt.Errorf("bootstrap targets controlplane nodes; node role is %q", node.Role)
	}

	args := []string{
		"--talosconfig", p.Talosconfig(),
		"--nodes", node.IP,
		"--endpoints", node.IP,
		"bootstrap",
	}
	runCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
	defer cancel()
	out, err := exec.CommandContext(runCtx, "talosctl", args...).CombinedOutput()
	if err != nil {
		msg := strings.TrimSpace(string(out))
		if strings.Contains(msg, "AlreadyExists") || strings.Contains(strings.ToLower(msg), "already bootstrapped") {
			// idempotent success
		} else {
			return fmt.Errorf("talosctl bootstrap failed: %s", msg)
		}
	}
	return waitForEtcd(ctx, p, node, timeout)
}

func waitForEtcd(ctx context.Context, p paths.Paths, node config.Node, timeout time.Duration) error {
	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}
		c, cancel := context.WithTimeout(ctx, 5*time.Second)
		out, err := exec.CommandContext(c, "talosctl",
			"--talosconfig", p.Talosconfig(),
			"--nodes", node.IP,
			"--endpoints", node.IP,
			"service", "etcd",
		).CombinedOutput()
		cancel()
		if err == nil && strings.Contains(string(out), "Running") && strings.Contains(string(out), "OK") {
			return nil
		}
		time.Sleep(3 * time.Second)
	}
	return fmt.Errorf("etcd did not become healthy on %s within %s", node.IP, timeout)
}

// FetchKubeconfig writes cluster kubeconfig to the nostos state directory.
func FetchKubeconfig(ctx context.Context, cfg *config.Config, p paths.Paths, node config.Node) error {
	if err := requireTalosctl(); err != nil {
		return err
	}
	if err := ensureTalosconfig(cfg, p); err != nil {
		return err
	}
	c, cancel := context.WithTimeout(ctx, 15*time.Second)
	defer cancel()
	out, err := exec.CommandContext(c, "talosctl",
		"--talosconfig", p.Talosconfig(),
		"--nodes", node.IP,
		"--endpoints", node.IP,
		"kubeconfig", "--force", p.Kubeconfig(),
	).CombinedOutput()
	if err != nil {
		return fmt.Errorf("kubeconfig fetch: %s", strings.TrimSpace(string(out)))
	}
	return nil
}

func ensureTalosconfig(cfg *config.Config, p paths.Paths) error {
	if validTalosconfig(p.Talosconfig()) {
		return nil
	}
	for _, candidate := range talosconfigCandidates(p) {
		if !validTalosconfig(candidate) {
			continue
		}
		body, err := os.ReadFile(candidate)
		if err != nil {
			return fmt.Errorf("read talosconfig %s: %w", candidate, err)
		}
		backends, err := secrets.BuildBackends(cfg)
		if err != nil {
			return err
		}
		rendered, err := secrets.ResolveTemplate(string(body), backends)
		if err != nil {
			return fmt.Errorf("resolve talosconfig %s: %w", candidate, err)
		}
		if err := os.MkdirAll(filepath.Dir(p.Talosconfig()), 0o755); err != nil {
			return err
		}
		if err := os.WriteFile(p.Talosconfig(), []byte(rendered), 0o600); err != nil {
			return err
		}
		return nil
	}
	return fmt.Errorf("no valid talosconfig found (checked %s)", strings.Join(talosconfigCandidates(p), ", "))
}

func talosconfigCandidates(p paths.Paths) []string {
	candidates := []string{
		filepath.Join(filepath.Dir(p.Root()), "talos", "talosconfig"),
		filepath.Join(p.Root(), "state", "talosconfig"),
	}
	if home, err := os.UserHomeDir(); err == nil && home != "" {
		candidates = append(candidates,
			filepath.Join(home, ".talos", "config"),
			filepath.Join(home, ".talos", "talosconfig"),
		)
	}
	return candidates
}

func validTalosconfig(path string) bool {
	body, err := os.ReadFile(path)
	if err != nil || len(strings.TrimSpace(string(body))) == 0 {
		return false
	}
	var tc struct {
		Context  string         `yaml:"context"`
		Contexts map[string]any `yaml:"contexts"`
	}
	if err := yaml.Unmarshal(body, &tc); err != nil {
		return false
	}
	return strings.TrimSpace(tc.Context) != "" && len(tc.Contexts) > 0
}

func requireTalosctl() error {
	if _, err := exec.LookPath("talosctl"); err != nil {
		return errors.New("talosctl not found; install: brew install siderolabs/tap/talosctl")
	}
	return nil
}

File: .submodules/nostos/src/nostos/cluster/bootstrap.py
"""talosctl bootstrap + wait-for-ready + fetch kubeconfig."""

from __future__ import annotations

import logging
import shutil
import subprocess
import time
from pathlib import Path

from ..config import NodeConfig, NostosConfig
from ..paths import Paths

log = logging.getLogger(__name__)


class BootstrapError(RuntimeError):
    pass


def _require_talosctl() -> None:
    if shutil.which("talosctl") is None:
        raise BootstrapError(
            "talosctl not found. Install: brew install siderolabs/tap/talosctl"
        )


def bootstrap_node(
    cfg: NostosConfig, paths: Paths, node: NodeConfig, *, timeout: float = 300.0
) -> None:
    """Run talosctl bootstrap; idempotent on already-bootstrapped cluster."""
    _require_talosctl()
    if node.role != "controlplane":
        raise BootstrapError(
            f"bootstrap targets controlplane nodes; node role is {node.role!r}"
        )

    if not paths.talosconfig.is_file():
        raise BootstrapError(
            f"no talosconfig at {paths.talosconfig}. Run `nostos config refresh` first."
        )

    ip = str(node.ip)
    log.info("bootstrapping etcd on %s", ip)

    try:
        result = subprocess.run(
            [
                "talosctl",
                "--talosconfig",
                str(paths.talosconfig),
                "--nodes",
                ip,
                "--endpoints",
                ip,
                "bootstrap",
            ],
            capture_output=True,
            text=True,
            timeout=30,
        )
    except subprocess.TimeoutExpired as e:
        raise BootstrapError(f"talosctl bootstrap timed out after 30s") from e

    if result.returncode != 0:
        # Idempotent-success: "AlreadyExists" from Talos means etcd is up.
        if "AlreadyExists" in result.stderr or "already bootstrapped" in result.stderr.lower():
            log.info("cluster already bootstrapped on %s — continuing", ip)
        else:
            raise BootstrapError(
                f"talosctl bootstrap failed: {result.stderr.strip() or result.stdout.strip()}"
            )

    wait_for_etcd(cfg, paths, node, timeout=timeout)


def wait_for_etcd(
    cfg: NostosConfig, paths: Paths, node: NodeConfig, *, timeout: float = 300.0
) -> None:
    """Poll talosctl service etcd until Running/OK or timeout."""
    _require_talosctl()
    ip = str(node.ip)
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        try:
            r = subprocess.run(
                [
                    "talosctl",
                    "--talosconfig",
                    str(paths.talosconfig),
                    "--nodes",
                    ip,
                    "--endpoints",
                    ip,
                    "service",
                    "etcd",
                ],
                capture_output=True,
                text=True,
                timeout=5,
            )
            if r.returncode == 0 and "Running" in r.stdout and "OK" in r.stdout:
                log.info("etcd healthy on %s", ip)
                return
        except subprocess.TimeoutExpired:
            pass
        time.sleep(3)
    raise BootstrapError(f"etcd did not become healthy on {ip} within {timeout}s")


def fetch_kubeconfig(paths: Paths, node: NodeConfig, *, force: bool = True) -> Path:
    """Save cluster kubeconfig to state/kubeconfig."""
    _require_talosctl()
    ip = str(node.ip)
    args = [
        "talosctl",
        "--talosconfig",
        str(paths.talosconfig),
        "--nodes",
        ip,
        "--endpoints",
        ip,
        "kubeconfig",
        str(paths.kubeconfig),
    ]
    if force:
        args.append("--force")
    try:
        subprocess.run(args, check=True, capture_output=True, text=True, timeout=15)
    except subprocess.CalledProcessError as e:
        raise BootstrapError(
            f"kubeconfig fetch failed: {e.stderr.strip() or e.stdout.strip()}"
        ) from e
    return paths.kubeconfig
